tls/extensions/s2n_server_renegotiation_info.c

```diff
@@ -102,6 +103,84 @@ static int s2n_renegotiation_info_recv(struct s2n_connection *conn, struct s2n_s
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc5746#3.5
+     *# This text applies if the connection's "secure_renegotiation" flag is
+     *# set to TRUE (if it is set to FALSE, see Section 4.2).
+     */
+    POSIX_ENSURE(conn->secure_renegotiation, S2N_ERR_NO_RENEGOTIATION);
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc5746#3.5
+     *# o  The client MUST then verify that the first half of the
+     *#    "renegotiated_connection" field is equal to the saved
+     *#    client_verify_data value, and the second half is equal to the
+     *#    saved server_verify_data value.  If they are not, the client MUST
+     *#    abort the handshake.
+     */
+
+    uint8_t renegotiated_connection_len = 0;
+    POSIX_GUARD(s2n_stuffer_read_uint8(extension, &renegotiated_connection_len));
+    POSIX_ENSURE(verify_data_len * 2 == renegotiated_connection_len, S2N_ERR_BAD_MESSAGE);
+
+    uint8_t *first_half = s2n_stuffer_raw_read(extension, verify_data_len);
+    POSIX_ENSURE_REF(first_half);
+    POSIX_ENSURE(s2n_constant_time_equals(first_half, conn->handshake.client_finished, verify_data_len),
+            S2N_ERR_BAD_MESSAGE);
+
+    uint8_t *second_half = s2n_stuffer_raw_read(extension, verify_data_len);
+    POSIX_ENSURE_REF(second_half);
+    POSIX_ENSURE(s2n_constant_time_equals(second_half, conn->handshake.server_finished, verify_data_len),
+            S2N_ERR_BAD_MESSAGE);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_renegotiation_info_recv(struct s2n_connection *conn, struct s2n_stuffer *extension)
+{
+    if (s2n_handshake_is_renegotiation(conn)) {
+        POSIX_GUARD(s2n_renegotiation_info_recv_renegotiation(conn, extension));
+    } else {
+        POSIX_GUARD(s2n_renegotiation_info_recv_initial(conn, extension));
+    }
+    return S2N_SUCCESS;
+}
+
+static int s2n_renegotiation_info_if_missing(struct s2n_connection *conn)
+{
+    POSIX_ENSURE_REF(conn);
+    if (s2n_handshake_is_renegotiation(conn)) {
+        /**
+         *= https://tools.ietf.org/rfc/rfc5746#3.5
+         *# o  When a ServerHello is received, the client MUST verify that the
+         *#    "renegotiation_info" extension is present; if it is not, the
+         *#    client MUST abort the handshake.
+         */
+        POSIX_BAIL(S2N_ERR_NO_RENEGOTIATION);
+    } else {
+        /**
+         *= https://tools.ietf.org/rfc/rfc5746#3.4
+         *# *  If the extension is not present, the server does not support
+         *#    secure renegotiation; set secure_renegotiation flag to FALSE.
+         *#    In this case, some clients may want to terminate the handshake
+         *#    instead of continuing; see Section 4.1 for discussion.
+         *
+         * We do not terminate the handshake, although missing messaging for secure
+         * renegotiation degrades server security.
+         *
+         * We could introduce an option to fail in this case in the future.
+         */
+        conn->secure_renegotiation = false;
```

Comment:
to be clear.. we dont allow reneg if this is not present right? since we have `POSIX_BAIL(S2N_ERR_NO_RENEGOTIATION);` above.

so why is there security degradation?

Reply:
That's a fun question :) It's worth reading the introduction to the secure renegotiation RFC: https://www.rfc-editor.org/rfc/rfc5746#section-1

Basically, the attack that prompted that RFC involved tricking a connection into thinking it WASN'T doing renegotiation, while a malicious actor secretly was doing renegotiation with it. So we expect the "secure renegotiation" signals-- either the extension or the cipher suite-- whether or not we support renegotiation, because even if we don't support renegotiation we have to indicate that this isn't a renegotiation handshake.

If a peer doesn't present the "secure renegotiation" signals, they either haven't implemented the defenses against that attack, or that attack is actually happening. But if we fail the handshake, we might be rejecting a legit connection. The RFC has a whole section on that choice: https://www.rfc-editor.org/rfc/rfc5746#section-4 I haven't changed the existing behavior though-- we've always accepted non-renegotiation handshakes without the "secure renegotiation" signals. I'd love to make that more strict since that RFC is now a decade old, but that might break existing customers and is out of scope.

This is a particularly bad example of the existence of an insecure feature compromising the rest of the protocol.